### ISSUE TITLE
Hangs and missing pages of data in multi-locale shared file write

### ISSUE THREAD

**Issue by psath** [chapel-lang/chapel#24462]
### Summary of Problem
  When writing to non-overlapping regions of arrays within a shared binary file on a parallel filesystem from multiple nodes, chunks of arrays are missing without throwing an error.
  These chunks appear to start at a page boundary and extend either to: another page boundary, the array boundary, or the boundary of the writing task's chunk of the array. The location and number of missing pages is non-deterministic, but requires some portion of work on at least a 2nd locale to appear.
  
  This issue was first observed when opening the write file once-per-task in R/W mode. The program reliably finishes, and sometimes produces a complete output.
  
  After discussion, it was suggested that the file be opened only once in the main task, and each distributed sub-task create a fileWriter from a shared ref to the file. This sometimes finishes, sometimes produces correct output, but sometimes completely hangs both nodes of the computation, with both nodes spinning the CPU at nearly 100%.
  
  Both GPFS and Qumulo have been used as the parallel filesystem, with similar results.
  
  This is blocking work on a scaling study of multi-node + multi-GPU Chapel for irregular workloads. (The bugs appear even if the GPU sublocale is unused, but I have not tried to reproduce without the GPU locale model.)
  
  Prior discussion was here: https://matrix.to/#/!PBYDSerrfYujeStENM:gitter.im/$gvfNNZvD0PdcWU5sQVjhaNnzvXNssqpXLEotXu6bnGc?via=gitter.im&via=matrix.org&via=envs.net
  
  ### Steps to Reproduce
  I've attached a stripped-down reproducer+makefile and the associated Slurm batch and wrapper scripts. Just remove all the `.txt` extensions that GitHub forces.
  
  Intended use is to run the code in producer mode once on a single locale to create a test file, and then run it multiple times in consumer mode with different locale and partition counts. (A minimum of 2 partitions and 2 locales will be necessary to trigger the bug). The size/density of the test file can be scaled if needed, but I've seen the bug show up on as small as 1MiB. See `batch.sh` for intended use.
  
  I've anonymized the code as much as I can, but the following variables may need to be provided
  - CHPL_LAUNCHER_ACCOUNT and CHPL_LAUNCHER_PARTITION in `modules.sh` (as well as the modules to fit your cluster)
  - CHPL_CUDA_PATH in `install_chapelGithub.sh` and `Makefile` to suit your cluster
  
  **Source Code:**
  `multi-locale-page-errors.chpl`
  
  **Compile command:**
  `make`
  
  **Execution command:**
Producer:
` ./mlpErrors --numLocales=1 --producer=true --targetFileLength=<amount in bytes> --filePrefix=<someStr>`  --> Produces `someStr.csr` 
Consumer:
`./mlpErrors --numLocales=2 --numParts=4 --filePrefix=<someStr>` --> partitioned-reads `someStr.csr` and partitioned-copies it to `someStr.out.csr`

  **Associated Future Test(s):**
- Additional parallel fileReader segfault discovered while generating the reproducer. If you uncomment line 212 and delete lines 209, 210, and 217, it will sometimes appear when `numLocales=2` and `numParts>=2`
  
  ### Configuration Information
  `install_chapelGithub.sh` provides the build configuration.
  
  - Output of `chpl --version`:
```
 warning: The prototype GPU support implies --no-checks. This may impact debuggability. To suppress this warning, compile with --no-checks explicitly
chpl version 1.34.0 pre-release (7efadb0)
  built with LLVM version 17.0.6
  available LLVM targets: amdgcn, r600, nvptx64, nvptx, aarch64_32, aarch64_be, aarch64, arm64_32, arm64, x86-64, x86
Copyright 2020-2024 Hewlett Packard Enterprise Development LP
Copyright 2004-2019 Cray Inc.
(See LICENSE file for more details)
```
  - Output of `$CHPL_HOME/util/printchplenv --anonymize`:
```
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: unknown
CHPL_LOCALE_MODEL: flat
CHPL_COMM: gasnet *
  CHPL_COMM_SUBSTRATE: udp *
  CHPL_GASNET_SEGMENT: everything
CHPL_TASKS: qthreads
CHPL_LAUNCHER: amudprun *
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: none
CHPL_HWLOC: bundled
CHPL_RE2: none
CHPL_LLVM: unset
CHPL_AUX_FILESYS: none
```
  - Back-end compiler and version, e.g. `gcc --version` or `clang --version`:
```
gcc (GCC) 11.2.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```
  - (For Cray systems only) Output of `module list`:
```
Currently Loaded Modules:
  1) shared                       9) binutils/2.37-GCCcore-11.2.0     17) SQLite/3.36-GCCcore-11.2.0
  2) slurm/slurm/23.02.7         10) bzip2/1.0.8-GCCcore-11.2.0       18) GMP/6.2.1-GCCcore-11.2.0
  3) apps                        11) OpenSSL/1.1                      19) libffi/3.4.2-GCCcore-11.2.0
  4) useful_scripts              12) cURL/7.78.0-GCCcore-11.2.0       20) CUDA/11.7.0
  5) DefaultModules              13) XZ/5.2.5-GCCcore-11.2.0          21) GCC/11.2.0
  6) site/infer/easybuild/setup  14) libarchive/3.5.1-GCCcore-11.2.0  22) CMake/3.22.1-GCCcore-11.2.0
  7) GCCcore/11.2.0              15) libreadline/8.1-GCCcore-11.2.0   23) ncurses/6.2-GCCcore-11.2.0
  8) zlib/1.2.11-GCCcore-11.2.0  16) Tcl/8.6.11-GCCcore-11.2.0        24) Python/3.9.6-GCCcore-11.2.0
```
  
[batch.sh.txt](https://github.com/chapel-lang/chapel/files/14364048/batch.sh.txt)
[install_chapelGithub.sh.txt](https://github.com/chapel-lang/chapel/files/14364049/install_chapelGithub.sh.txt)
[Makefile.txt](https://github.com/chapel-lang/chapel/files/14364050/Makefile.txt)
[modules.sh.txt](https://github.com/chapel-lang/chapel/files/14364051/modules.sh.txt)
[multi-locale-page-errors.chpl.txt](https://github.com/chapel-lang/chapel/files/14376245/multi-locale-page-errors.chpl.txt)
[wrapper.sh.txt](https://github.com/chapel-lang/chapel/files/14364053/wrapper.sh.txt)

**Comment by psath** [comment]
Per @jeremiah-corrado 's suggestion, I added `-sIOSkipBufferingForLargeOps=false` to the Makefile. This seems to have mitigated the hangs but I am now encountering a new segfault here:
```
[0] Thread 6 (Thread 0x2aaab5000700 (LWP 234347)):
[0] #0  0x00002aaaad45c659 in waitpid () from /lib64/libc.so.6
[0] #1  0x00002aaaad3d9f62 in do_system () from /lib64/libc.so.6
[0] #2  0x00002aaaad3da311 in system () from /lib64/libc.so.6
[0] #3  0x000055555598913a in gasneti_system_redirected ()
[0] #4  0x0000555555988cd0 in gasneti_bt_gdb ()
[0] #5  0x0000555555982cb0 in gasneti_print_backtrace ()
[0] #6  0x0000555555573043 in gasneti_defaultSignalHandler ()
[0] #7  <signal handler called>
[0] #8  0x00002aaaad4ebd29 in __memcpy_ssse3_back () from /lib64/libc.so.6
[0] #9  0x0000555555906d6a in qbuffer_copyin ()
[0] #10 0x00005555558ec019 in _qio_slow_write ()
[0] #11 0x00005555558a1d41 in qio_channel_write_amt(int, qio_channel_s*, void const*, long) ()
[0] #12 0x000055555583c260 in writeBinary_chpl ()
[0] #13 0x00005555557a1c8c in writeBulkElements_chpl2 ()
[0] #14 0x00005555556339bf in chpl_serialReadWriteRectangularHelper6 ()
[0] #15 0x000055555562ffbf in chpl_serialReadWriteRectangular8 ()
[0] #16 0x000055555562f10c in chpl_serialReadWriteRectangular5 ()
[0] #17 0x000055555562e581 in dsiSerialReadWrite_chpl7 ()
[0] #18 0x0000555555666821 in dsiSerialWrite_chpl8 ()
[0] #19 0x0000555555596d1c in writeThis_chpl3 ()
[0] #20 0x0000555555597f40 in serialize_chpl5 ()
[0] #21 0x000055555579cefd in serializeValue_chpl17 ()
[0] #22 0x00005555557fc770 in _serializeOne_chpl ()
[0] #23 0x0000555555850f80 in on_fn_chpl220 ()
[0] #24 0x000055555587c782 in serialize_chpl10 ()
[0] #25 0x000055555579c478 in serializeValue_chpl15 ()
[0] #26 0x00005555557fcedf in _serializeOne_chpl10 ()
[0] #27 0x000055555584bf79 in on_fn_chpl213 ()
[0] #28 0x00005555558572ec in wrapon_fn_chpl218 ()
[0] #29 0x00005555558bd0da in fork_wrapper ()
[0] #30 0x00005555558b8b33 in chapel_wrapper ()
[0] #31 0x000055555594c312 in qthread_wrapper ()
[0] #32 0x0000000000000000 in ?? ()
```

Moreover, the non-faulting application runs now have repeating streaks of bad/misaligned data in the damaged sections of file, where I previously would typically see zero-fill. (i.e. the `offsets` and `weights` arrays. `indices` in the middle previously had what looks like patterned, but misaligned data.)

**Comment by jeremiah-corrado** [comment]
When running the consumer part of your reproducer on main (not using the GPU locale model) with multiple locales, I'm seeing the following runtime error:
```
mlpErrors.chpl:82: error: references to remote data cannot be passed to external routines like 'c_pointer_return'
```

Where line 82 is:
```chapel
try! writer.write(offsets[0..#oSize]);
```

I still need to dig into this a little further, but I'm guessing that the array slice created by `offsets[0..#oSize]` is somehow not local in all cases, which is causing the warning to be emitted when `writeBinary` creates a c-pointer to the slice's first element (#11 in your stack trace above).

Have you seen this error come up in your runs @psath? Are you running with checks enabled?

**Comment by psath** [comment]
IIRC the GPU sublocale disables checks, so that hasn't been raised for me.

Based on some of the testing I was doing last week, what you're getting at does worry me.

The `CSR_part` object that owns `offset[0..#oSize]` should be local to the locale that owns that partition, it may not be the main locale. However, if I recall, when a `fileWriter` is opened on some locale that differs from the `file` it is created from, it run's on `file`'s locale instead. (However, this is the pattern that was recommended to me in the gitter thread).

But if the `CSR_part` is owned by locale 1, and its `serialize` method is running on locale 0 because that's where the `fileWriter` is technically running, then yes `offsets` would be necessarily be non-local w.r.t. where the fileWriter is for any locale other than 0.

**Comment by psath** [comment]
So, noticing this line now specifies `local` https://github.com/chapel-lang/chapel/blob/1349d09640202a4e2ae87431d8ddd1349d69aa4c/modules/standard/IO.chpl#L8461

It makes me think my original formulation might've been the correct approach:
Essentially: every partition has its own `file` pointing to the same file on the parallel filesystem, which means the `fileWriter` is local to whichever locale that partition is running on (obeying the [new-ish?] restriction on the array form of `writeBinary`, since `serialize` will be called on the same locale that owns `CSR_part.offsets`). And then lean on the parallel filesystem to handle writes from different nodes that happen to be to the same page.

To that end, I modified the consumer as attached, but kept the `-sIOSkipBufferingForLargeOps=false` flag in the makefile.
And lo and behold, all the tests came up PASS.

Going to see if that flag helps the main code as originally written.

[multi-locale-page-errors.chpl.txt](https://github.com/chapel-lang/chapel/files/14378212/multi-locale-page-errors.chpl.txt)

**Comment by psath** [comment]
Ok, still sparingly seeing a few pages of zero-fill from the reproducer at problem sizes larger than 64 MB, but at a substantially-reduced rate

**Comment by jeremiah-corrado** [comment]
I was just thinking along similar lines before I saw your message.  And I was able to remove the non-local access with the following modification to `serialize`:

```chapel
 on writer._home {
  const olocal = offsets[0..#oSize],
        ilocal = indices,
        wlocal = weights;
  
  try! writer.seek((startOfOffsets+(partStart*numBytes(offsets.eltType)))..#(oSize*numBytes(offsets.eltType)));
  try! writer.write(olocal);
  var iwSize : int(64) = partEndIdx-partStartIdx;
  try! writer.seek((startOfIndices+(partStartIdx*numBytes(indices.eltType)))..#(iwSize*numBytes(indices.eltType)));
  try! writer.write(ilocal);
  try! writer.seek((startOfWeights+(partStartIdx*numBytes(weights.eltType)))..#(iwSize*numBytes(weights.eltType)));
  try! writer.write(wlocal);
}
```

But this strategy does of course require an array copy. So, as you said, a formulation with separate files on each locale might be the more appropriate pattern. That being said, I'm not sure what the implications of:

> And then lean on the parallel filesystem to handle writes from different nodes that happen to be to the same page.

are with Chapel's I/O runtime design. @mppf, do you know if the I/O module supports concurrent writes to the same file from multiple `file` objects?

**Comment by jeremiah-corrado** [comment]
Separately, I'll note that the lack of an error message / warning about the non-local access is something we want to address. The `writeBinary` method does have a check for this that should fire even with `--checks` turned off, but it was bypassed in this case for somewhat extraneous reasons. I'll open a separate issue about that.

**Comment by mppf** [comment]
> @mppf, do you know if the I/O module supports concurrent writes to the same file from multiple `file` objects?

Yes that should work OK, provided that, if you are writing, you are writing to different regions (ideally with disjoint `region` arguments when constructing the fileWriters). Of course there might be bugs here.

> But If I have to _first_ communicate it to the main locale, which _then_ communicates it to the parallel file system, that introduces an additional large network comm _and_ a serial bottleneck which is a little incongruous with a parallel [and distributed]-first language.

That's an interesting and good point. I have been historically thinking that implementing #7953 would address many of these concerns. However I think you are bringing up something different from what that issue would solve.

The current way to do this is by opening a `file` object on each locale rather than using a global `file`.

If we want to, we could implement it so that different `fileWriter` / `fileReader`s on different locales to the same file that is available on a network filesystem actually work with different files under the hood. (The implementation would need to detect that it's a shared filesystem & create one localized `file` per locale). Of course, using a different file per-locale can achieve the same thing today, it's just arguably a style that shouldn't be required to get the optimal I/O pattern in this context. Arguably, having one `file` and many `fileWriter`s is a more natural style for a Chapel program. Anyway, if you think this idea is worthwhile, we should make an issue about it.

**Comment by psath** [comment]
> That's an interesting and good point. I have been historically thinking that implementing #7953 would address many of these concerns. 

If I'm understanding #7953 correctly... I think that's a very appealing case when one node has sufficient/fast-enough storage and either 1) no parallel filesystem exists or 2) the parallel filesystem isn't particularly fast (my case: Qumulo is in a different building, and I'm not exactly sure where the GPFS is physically located relative to the GPU nodes.)

> However I think you are bringing up something different from what that issue would solve.
>
> The current way to do this is by opening a `file` object on each locale rather than using a global `file`.
> 
> If we want to, we could implement it so that different `fileWriter` / `fileReader`s on different locales to the same file that is available on a network filesystem actually work with different files under the hood. (The implementation would need to detect that it's a shared filesystem & create one localized `file` per locale). Of course, using a different file per-locale can achieve the same thing today, it's just arguably a style that shouldn't be required to get the optimal I/O pattern in this context. Arguably, having one `file` and many `fileWriter`s is a more natural style for a Chapel program. Anyway, if you think this idea is worthwhile, we should make an issue about it.

Uhhhh if there was a way for these shadow-files to be mapped to a fast local SSD as a buffer before they get flushed to the parallel filesystem, maybe. But that seems like it would be more FS or OS scope than Chapel's job? I guess I'm balking on adding an additional layer of runtime complexity when I'm already struggling to get globally-consistent writes to work.

**Comment by jeremiah-corrado** [comment]
> Ok, still sparingly seeing a few pages of zero-fill from the reproducer at problem sizes larger than 64 MB, but at a substantially-reduced rate. And I've yet to see any evidence of the patterned/garbage data from previous designs, strictly zero-fill.

Okay, glad to see things have improved! I'll keep digging into this and see if I can find the cause of the missing pages.

**Comment by psath** [comment]
> Ok, still sparingly seeing a few pages of zero-fill from the reproducer at problem sizes larger than 64 MB, but at a substantially-reduced rate. And I've yet to see any evidence of the patterned/garbage data from previous designs, strictly zero-fill.

So I did a little extra instrumentation of `CSR_part.serialize` to see how the remaining zero-filled partial pages aligned with the actual array partitions.

I added these lines
```diff
         var startOfWeights = startOfIndices + (numEdges)*numBytes(indices.eltType);
+        writeln("partID = ", partID, ", startOfOffsets = ", startOfOffsets, ", startOfIndices = ", startOfIndices, ", startOfWeights = ", startOfWeights, ", endOfWeights = ", startOfWeights +(numEdges)*numBytes(weights.eltType));
```
```diff
         try! writer.write(weights);
+        writeln("partID = ", partID, ", offsetSeekRange = ", ((startOfOffsets+(partStart*numBytes(offsets.eltType)))..#(oSize*numBytes(offsets.eltType))), ", indicesSeekRange = ", ((startOfIndices+(partStartIdx*numBytes(indices.eltType)))..#(iwSize*numBytes(indices.eltType))), ", weightsSeekRange = ", ((startOfWeights+(partStartIdx*numBytes(weights.eltType)))..#(iwSize*numBytes(weights.eltType))));
```

And what it looks like is that, when there is a problem it's consistently the last page in _an **even**_ task's chunk of the array(s). (It's curious that it's only the boundaries between even and odd tasks, not odd and even... i.e the 0-->1, 2-->3, and 6-->7 task boundaries. NEVER the 1-->2 or 5-->6)

This would be consistent with 2 possible options I can think of offhand:
1) The even locale _silently gives up_ on a write when it is notified of a dirty page that the odd locale modified. 
2) The odd locale _silently overwrites_ a prior write from the even locale.

Both could be possible outcomes of a race between locale 1 reading the start of its chunk, locale 0 reading the end of its chunk, locale 1 writing the start of its chunk, and locale 0 writing the end of its chunk, but then locale 0 has a conflict because the page has been modified by locale 1.

I would expect the FS/OS to raise an error about a dirty page conflict or something and allow the application to catch and handle the conflicting write however it wished (halt, silent overwrite, reread and replay the write..) The desired behavior, since the written array ranges are non-overlapping, would be to re-read/replay. The page conflict won't happen again, as each task only writes the arrays once, but in the general case you'd probably need to keep re-read/replaying until success right?

I do also see multiple cases with `numParts>=4` where there is an additional problem with the first page (`partID=0`'s start of `offsets`) which would be consistent with the conflicting re-write of the file's shared header. That is clearly redundant w/ the main task's `.cw` initialization of the file's header, so the `super.serialize` can be eliminated entirely.

That's it for me today, and I'll be fairly tied up tomorrow, but it feels like we're getting closer, so thanks folks!

**Comment by psath** [comment]
Sigh, should've double-checked it earlier, but I suspect this is a cluster setup issue more than a Chapel one.

Even though the cluster filesystem is GPFS, it's exported to the work nodes through NFSv4, which has a weaker (close-to-open) consistency model than GPFS is capable of (locked byte-ranges).

Not sure I'll get to test it today, but a feasible workaround would be a simple lock to ensure tasks with adjacent write regions never co-open the file. If that works, then I'll see if I can back off the `-sIOSkipBufferingForLargeOps=false` flag.

**Comment by mppf** [comment]
@psath - thanks for continuing to wrestle with it.

Maybe it's a dumb idea, but in principle you should be able to open a new file in the `serialize` method to write the data at a particular offset. You could use e.g. `var localFile = open(fileWriter.getFile().path, ioMode.rw)` after PR #24364 (which solved your issue #23066). In principle, this could solve both of these problems:
 1. Not wanting to have extra copying of the data on the way to the network filesystem (avoids it by having the file/fileWriter be local to the operation)
 2. Fitting into the NFS close-to-open consistency (since you can open and then close separate files for each region you are writing to in parallel). This part I am less sure of, because I'm not confident I understand close-to-open right now. But maybe it would help. Perhaps NFS will be happier with this pattern if you open a separate file for each of the `seek` calls.

(FWIW, I haven't made up my mind whether I think it's better to avoid `seek` in `serialize`, but I also think that's mostly irrelevant at this point. But, I'm pretty sure that `serialize` calls for subsequent objects (say if you did `write(part1, part2)`) would start wherever the `seek` left things. That's not normally what you want, but you could address it by having the `serialize` finish up by `seek`ing to wherever the next object should start. I think this is probably irrelevant for your test program because AFAIK it's only writing one CSR_part per `write`. Anyway, IMO your code is probably better off using a different method from `serialize` for seeking and reading/writing.)

> EDIT: One last option that comes to mind, is that the `.cw` initialization of the file only allocates the header's worth of space; the seek-to-extend is performed by each task. So I could potentially understand how the tasks might not think they needed to check what was currently on disk if the last time they checked the file was shorter than where they're seeking to. Next time I have a chance to look at it I can try pre-allocating the entire file-length.

This reminds me of some strange code in the Chapel runtime implementation for getting the file to the right size when writing with `mmap` (because we have to extend the file by a whole page in order to mmap it, but then it might end up too long and need truncating) but I suspect you wouldn't be running that today (since IIRC we turned off using mmap by default to avoid portability problems). Either way, if it had to do with that code, using the `region` argument should keep it from doing anything too strange.

Anyway I agree that sizing the file before you start cuts out some variables. (AFAIK, the OS will zero-fill it without actually using any storage space, so it doesn't actually reserve disk space, but it might make it clearer to the OS / FS what you are doing).

**Comment by mppf** [comment]
One more parenthetical, I think these NFS issues are a big part of the reason that some array file formats (e.g. [Zarr](https://zarr-specs.readthedocs.io/en/latest/v3/core/v3.0.html#introduction)) use different files for different "chunks" of data. In fact the situation is probably worse with something like S3 (where, IIRC, you can't even modify a file ever).

**Comment by psath** [comment]
So I've got a modified version of the test running with a new user-space `safeWrite` function that's clean up to 256MB so far. (yesterday we were hitting errors by 64MB.) Fingers crossed.

This makes me rethink my interest in a variation on your idea here:

>If we want to, we could implement it so that different fileWriter / fileReaders on different locales to the same file that is available on a network filesystem actually work with different files under the hood. (The implementation would need to detect that it's a shared filesystem & create one localized file per locale). Of course, using a different file per-locale can achieve the same thing today, it's just arguably a style that shouldn't be required to get the optimal I/O pattern in this context. Arguably, having one file and many fileWriters is a more natural style for a Chapel program. Anyway, if you think this idea is worthwhile, we should make an issue about it.

Could [/should] Chapel support a concept of a "network file"? You open the file once in the main task, but `fileWriter`s on other locales could "own" specific byte/page ranges (according to different network file consistency models), allowing them to *mostly* parallel write to the network filesystem. The owner could default to whoever first owned that page/byte-range without closing it, and transfer ownership to the oldest open fileWriter as tasks wrap up and flush. Any writes to non-owned byte/page-ranges would be communicated to the owning locale to write.

In my case this would mean the long extents of pages that strictly belong to one task would be owned by that task's `on` locale, but the contested pages would simply be owned by the locale of whichever task opened them first. Most writes would go straight to the NFS, but only contested pages would go through the slower two-hop path to owner and then to the NFS.

**Comment by psath** [comment]
Sure enough, all green up to 2GB files. It does not need the `-sIOSkipBufferingForLargeOps=false` flag.

Here's how I set up the `safeWrite` function. I'll have to restrict `numLocales` and `numParts` to both be even and I might lose a hair of throughput, but it works
```diff
+  proc safeWrite(ref lock : atomic int, ref data : CSR_part) {
+    var myID = here.id;
+    myID = if (myID == 0) then Locales.size else myID; //Replace locale 0's reserved value
+    myID = if ((myID & 1) != 0) then -myID else myID; //Replace odd locales' values
+    var cur = lock.read();
+    while (true) {
+      if (cur == 0 || sgn(cur) == sgn(myID)) { //Should try to acquire
+        //Use a sum so co-occupiers can acquire/release in any order
+        var newVal = cur+myID;
+        if(lock.compareExchange(cur, newVal)) { //Success, cur was still valid
+          writeln(here.id, " acquired writeLock with val ", newVal);
+          cur = newVal;
+          //Do write
+          var myoutFile = try! IO.open(filePrefix + ".out.csr", ioMode.rw);
+          var myWriter = try! myoutFile.writer(serializer = new binarySerializer(endian = ioendian.little, _structured = false), locking = false, hints = IO.ioHintSet.sequential);
+          try! myWriter.write(data);
+          try! myWriter.flush();
+          try! myWriter.close();
+          try! myoutFile.fsync();
+          try! myoutFile.close();
+          //Release
+          while (true) { //Other co-occupiers may modify cur, so keep trying
+            newVal = cur-myID;
+            if (lock.compareExchange(cur, newVal)) {
+              writeln(here.id, " released writeLock with val ", newVal);
+              return; //Success
+            }
+          }
+        }
+      }
+      //Occupied by possibly-adjacent task(s) or failed to acquire
+      Time.sleep(safeWriteSleepSec);
+      cur = lock.read();
+    }
+  }
```

Complete reproducer attached for posterity. As soon as I can test this approach on the full app, I will close as completed but feel free to open others about the other segfaults, ideas, etc. raised in the thread.

[multi-locale-page-errors.chpl.txt](https://github.com/chapel-lang/chapel/files/14388594/multi-locale-page-errors.chpl.txt)

**Comment by psath** [comment]
Works in the full app as well.

**Comment by bradcray** [comment]
Catching up after vacation, I'm happy to see this resolved!  I'm also wondering whether there are issues we should spin off from it, such as:

* Is there more we should be doing for users of the GPU locale model to deal with checks being disabled?
* Do we want to capture Paul's idea of adding a "network file" concept to the library?
* Are there other things we could've / should've done here to make the errors that were hit along the way more evident / less likely to trip up other users?

**Comment by jeremiah-corrado** [comment]
> Are there other things we could've / should've done here to make the errors that were hit along the way more evident

#24480 was opened asking if an error should be thrown when attempting to write a non-local array using the binary serializer (where the non-local access was one of the main issues Paul was running into after he switched to using a single global file).

**Comment by mppf** [comment]
I've gone ahead and created issue #24499 about the "network file" idea.
